core/src/main/java/io/grpc/util/OutlierDetectionLoadBalancer.java

```diff
@@ -0,0 +1,1001 @@
+
+        // If the map has an entry for the new address, we associate this subchannel with it.
+        EquivalentAddressGroup newAddress = Iterables.getOnlyElement(addresses);
+        if (trackerMap.containsKey(newAddress)) {
+          AddressTracker tracker = trackerMap.get(newAddress);
+          tracker.addSubchannel(this);
+        }
+      } else if (getAllAddresses().size() == 1 && addresses.size() > 1) {
+        // We go from a single address to having multiple, making this subchannel uneligible for
+        // outlier detection. Remove it from all trackers and reset the call counters of all the
+        // associated trackers.
+        // Remove the current subchannel from the old address it is associated with in the map.
+        if (trackerMap.containsKey(getAddresses())) {
+          AddressTracker tracker = trackerMap.get(getAddresses());
+          tracker.removeSubchannel(this);
+          tracker.clearCallCounters();
+        }
+      } else if (getAllAddresses().size() > 1 && addresses.size() == 1) {
+        // We go from, previously uneligble, multiple address mode to a single address. If the map
+        // has an entry for the new address, we associate this subchannel with it.
+        EquivalentAddressGroup eag = Iterables.getOnlyElement(addresses);
+        if (trackerMap.containsKey(eag)) {
+          AddressTracker tracker = trackerMap.get(eag);
+          tracker.addSubchannel(this);
+        }
+      }
+
+      // We could also have multiple addresses and get an update for multiple new ones. This is
+      // a no-op as we will just continue to ignore multiple address subchannels.
+
+      delegate.updateAddresses(addresses);
+    }
+
+    /**
+     * If the {@link Subchannel} is considered for outlier detection the associated {@link
+     * AddressTracker} should be set.
+     */
+    void setEquivalentAddressGroupInfo(AddressTracker eagInfo) {
+      this.eagInfo = eagInfo;
+    }
+
+    void eject() {
+      ejected = true;
+      subchannelStateListener.onSubchannelState(
+          ConnectivityStateInfo.forTransientFailure(Status.UNAVAILABLE));
+    }
+
+    void uneject() {
+      ejected = false;
+      if (lastSubchannelState != null) {
+        subchannelStateListener.onSubchannelState(lastSubchannelState);
+      }
+    }
+
+    boolean isEjected() {
+      return ejected;
+    }
+
+    @Override
+    protected Subchannel delegate() {
+      return delegate;
+    }
+
+    /**
+     * Wraps the actual listener so that state changes from the actual one can be intercepted.
+     */
+    class OutlierDetectionSubchannelStateListener implements SubchannelStateListener {
+
+      private final SubchannelStateListener delegate;
+
+      OutlierDetectionSubchannelStateListener(SubchannelStateListener delegate) {
+        this.delegate = delegate;
+      }
+
+      @Override
+      public void onSubchannelState(ConnectivityStateInfo newState) {
+        if (!ejected) {
+          lastSubchannelState = newState;
```

Comment:
I had first interpreted the spec to mean we need to save the last updated status since the ejection, but it is just "the latest". So you are right, I changed it.